zaza/openstack/charm_tests/vault/utils.py

```diff
@@ -280,6 +288,35 @@ def ensure_secret_backend(client):
         pass
 
 
+def wait_for_ready_all(clients):
+    """Return once vault is ready (for all clients).
+
+    Raise an error if it's not ready within a reasonable time.
+
+    :param clients: Clients to use to talk to vault
+    :type clients: List[CharmVaultClient]
+    """
+    for client in clients:
+        wait_for_ready(client)
+
+
+@tenacity.retry(
+    reraise=True,
+    wait=tenacity.wait_fixed(10),
+    stop=tenacity.stop_after_attempt(20))
+def wait_for_ready(client):
+    """Return once vault is ready.
```

Comment:
@UtkarshBhatthere if you were being really picky ( :) ) then I'd actually suggest a rename of the function.  If you look at what it does:

1. It checks to see if the client is ready.
2. It repeats that check a few times after 200 seconds (which seems like a long time to me).
3. It raises an exception.

I'd probably rename the function to `wait_until_ready`, and instead of raising `Exception`, I'd raise `AssertionError`.

Also, the docstring ought to contain the exception that may be raised and under what circumstances.  I'd also move the retry code into the function using a `with` statement; after all the function is supposed to wait until it's ready not just check and bail out; the wrap-around of tenacity makes that less obvious; moving tenacity *into* the function makes it clear that the function waits until it's ready for a certain during and then fails.

Reply:
That's even better! As a general rule I stay away from using terms which overlap with syntax (return, yield etc) since they make it a mix bag. Also +1 on AssertionError suggestion, catching Exception directly can result in unexpected errors sneak past quietly. 
PS: I was being picky :)

Reply:
> Thanks for your comments, I'll make updates as necessary. :)
> 
> > It repeats that check a few times after 200 seconds (which seems like a long time to me).
> 
> Maybe I'm not interpreting the tenacity rules correctly? I `wait_fixed(10), stop_after_attempt(20)` means it will check every 10 seconds, and stop after 20 checks.

Yup, that's correct, and 200 seconds if all the retries are exhausted.  Which seems a long time for the vault to be initialised.  In practice it'll probably be "a few seconds" or "never", but I could be wrong about that.